### ISSUE TITLE
Error 20001: StarIO10IllegalDeviceStateError

### ISSUE THREAD

**Issue by joarwilk** [star-micronics/react-native-star-io10#115]
## Description
When discovering, it fails with error 2001: StarIO10IllegalDeviceStateError.

When shipping a batch of, say, 5 terminals, this can happen on 1-5 of those. It seems to be random.

How can we debug this invalid state?

## Your device where the bug occurs 
 - Device: 
   Castle S1F2
 - OS:
   Verifone / Adyen, not sure about version

## Your printer
- Model Names:
   mC-Print3
- Firmware Version:
   Unsure
- Interface:
   LAN

## To Reproduce
Start discovery. An error is caught with error code 2001 and description StarIO10IllegalDeviceStateError.
Sometimes its error 2000, with description "y"

## Expected behavior
Printers are found

**Comment by can-miki** [comment]
I would like to ask you a few questions to confirm the situation.

My understanding is that there are 5 terminals (Castle S1F2) connected to 1 printer (mC-Print3) via LAN. Is that so?
What API do you use for discovery?

Also, regarding [StarIO10IllegalDeviceStateError](https://star-m.jp/products/s_print/sdk/react-native-star-io10/manual/en/api-reference/star-io10-illegal-device-state-error/index.html) I would like to check the following:
 - Is there a problem with your network equipment?
 - Is the printer turned on?

The flow when connecting the printer and device is "PortOpen → Status acquisition → PortClose".
 - Is the application configured to communicate between the device and the printer after "PortClose"?

**Comment by joarwilk** [comment]
> My understanding is that there are 5 terminals (Castle S1F2) connected to 1 printer (mC-Print3) via LAN. Is that so?
> What API do you use for discovery?
Correct, but not connected at the same time. For any action they open, <do action>, and close.

We use this library to discover. Using `starDeviceDiscoveryManagerFactory.create([InterfaceType.Lan]);`

> Is there a problem with your network equipment?
> Is the printer turned on?

It works for some terminals and some not in the same network. This is consistent across different venues so it doesnt seem to be a local thing.

> Is the application configured to communicate between the device and the printer after "PortClose"?

We dont even get to that state. It simply doesnt return any printers during discovery.

**Comment by nuzayets** [comment]
We're seeing this with 'Network Unavailable' even when the network is available (we observe traffic to our backend succeed, at the same time as this error). We see it across Android 12 and 13, on the latest version of star-io10. The devices are Lenovo YT-J706F and Lenovo TB125FU. The printer is TSP100iii on Bluetooth but this occurs during discovery. We also began seeing it in May '24, at the time on version 1.3.0. 

The error is persistent. So far we have not found a solution.

**Comment by can-miki** [comment]
@joarwilk @nuzayets 
We apologize for the long wait.
Error 2001 is [Network Unavailable](https://github.com/star-micronics/react-native-star-io10/blob/42c87e74d5f555bf677a6d80c4cbb7a6c56b4d07/src/StarIO10ErrorCode.ts).
This SDK ver.1.7.0 fixed an issue where StarIO10IllegalDeviceStateError(message: "Network Unavailable.") could occasionally be thrown even though the device is actually connected to the network.
Please update and see if the problem is resolved.

Thank you.

**Comment by nuzayets** [comment]
The fix there said it's only for iOS. We're on Android 12 & 13, no effect there. I did find a workaround: if we get Network Unavailable we scan without LAN and that works. Of course that will only work if people don't genuinely need LAN printers -- lucky for us that scenario hasn't happened since we don't officially support LAN printers (though some customers do use them). 

My thinking was their network is blocking UDP multicast or something since that's usually how printer autodiscovery works (for normal desktop printers anyway), it does seem to be specific to a customer location and we can't repro it in-house. Since the workaround (not scanning LAN if we get Network Unavailable) works, we haven't put resources into reproducing it in the lab as of yet.

**Comment by can-miki** [comment]
We apologize for the long wait.
@nuzayets 
Thank you for sharing the information. 

@joarwilk 
I prepared 1 printer and 4 Android devices and checked the [discovery](https://star-m.jp/products/s_print/sdk/react-native-star-io10/manual/en/searchPrinter.html) function and [printing](https://star-m.jp/products/s_print/sdk/react-native-star-io10/manual/ja/api-reference.html), but this error(StarIO10IllegalDeviceStateError) did not occur.
Please let me know if you have any comments or steps to reproduce.

Printer: MCP31LB (F/W ver.3.81)
Interface: LAN

Device:
Pixel4a (Android version: 13)
Pixel6 (Android version: 14)
Pixel Tablet (Android version: 14)
Pixel8a (Android version: 15)

StarXpand SDK for ReactNative: 1.8.0

> info Fetching system and libraries information...
(node:66813) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
System:
  OS: macOS 14.6.1
  CPU: (8) arm64 Apple M2
  Memory: 121.78 MB / 8.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 22.11.0
    path: /usr/local/bin/node
  Yarn:
    version: 1.22.22
    path: /usr/local/bin/yarn
  npm:
    version: 10.9.0
    path: /usr/local/bin/npm
  Watchman:
    version: 2024.07.01.00
    path: /usr/local/bin/watchman
Managers:
  CocoaPods:
    version: 1.15.2
    path: /usr/local/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 24.1
      - iOS 18.1
      - macOS 15.1
      - tvOS 18.1
      - visionOS 2.1
      - watchOS 11.1
  Android SDK: Not Found
IDEs:
  Android Studio: 2024.1 AI-241.18034.62.2412.12266719
  Xcode:
    version: 16.1/16B40
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.11
    path: /usr/bin/javac
  Ruby:
    version: 2.6.10
    path: /usr/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.3.1
    wanted: 18.3.1
  react-native:
    installed: 0.75.4
    wanted: 0.75.4
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: Not found
  newArchEnabled: false

**Comment by nuzayets** [comment]
Yeah we couldn't reproduce it either. It happened at a customer site reliably even after switching hardware (both printer and Android device). We don't control their network so it's entirely possible something on their network was causing the LAN discovery to throw. There is likely a bug in handling unexpected behaviour in LAN discovery but I don't have any ideas.

**Comment by can-miki** [comment]
Since this does not occur in the lab, there may be a problem with the end-user's network environment.

[As shown previously](https://github.com/star-micronics/react-native-star-io10/issues/115#issuecomment-2242582138), error 2001 is NetworkUnavailable.
[NetworkUnavailable means "The network function of the host device cannot be used."](https://star-m.jp/products/s_print/sdk/react-native-star-io10/manual/en/api-reference/star-io10-error-code/index.html), so it's likely that the host terminal is not connected to the network and an error has occurred.
In other words, from the printer's perspective, the host device is disconnected from the network. (For example, the device's Wi-Fi may be turned off.)

If you have any comments please let us know.

**Comment by ashleyww93** [comment]
We have now seen this too. It was an event in a conference hall (EUROCIS) today (setup day)

The device had network connection, and the printer could be accessed via the the IP in the browser.

**Comment by can-miki** [comment]
Thank you for your inquiry. So far, we have not been able to reproduce the issue in our lab, so we would appreciate it if you could provide detailed reproduction steps.  
Filling out a bug report would be helpful.  
Please also share information about the printer you are using and its settings ([a Self-Printing](https://star-m.jp/products/s_print/mcprint3/manual/en/basicOperations/testPrint.htm) would be helpful), as well as details about the Android device and the React Native environment if available.

**Comment by dgreasi** [comment]
@can-miki I have gathered information related to the issue. The devices that this occurred are the following:

1. Samsung SM-G975F - Android 12
2. Samsung SM-A145R - Android 15
3. iPhone 7 Plus - iOS 15.8.5
4. Samsung SM-S931B - Android 16
5. iPhone XS Max - iOS 18.7.1

The error is the same for all cases: `{"name":"StarIO10IllegalDeviceStateError","_errorCode":2001}`

Here are the printer settings.

![Image](https://github.com/user-attachments/assets/24508159-7d74-42f2-9c6f-6d6df6a87a23)
![Image](https://github.com/user-attachments/assets/8acd8321-71c5-43dd-aef6-0eb82cbb96b8)
![Image](https://github.com/user-attachments/assets/c76563d6-d239-4819-8742-813cf1739b17)
![Image](https://github.com/user-attachments/assets/a81ef832-257e-4077-a8ce-12f59e958862)
![Image](https://github.com/user-attachments/assets/732b57bd-7f16-4aca-a23c-f845bd545a0b)
![Image](https://github.com/user-attachments/assets/c599570e-b07b-4f05-8145-b2e39319ff7b)

**Comment by bandit-ibayashi** [comment]
We are continuing to investigate cases where the `StarIO10IllegalDeviceStateError.NetworkUnavailable` occurs, but we have not been able to reproduce it in our environment, so we have not yet identified the specific conditions.

@dgreasi 
Thank you for sharing the photos of the Self-Printing. This confirms that the printer is on a Class C local network and has been assigned an IP address. When this issue occurs, is the IP address of the host terminal correctly assigned? We believe it should be, but since this exception can occur if the IP address is not properly assigned, we would like to confirm it just to be sure.

**Comment by joarwilk** [comment]
I had this happen on a device at the office. I then upgraded to the latest version of the library, and it worked. This was on iOS.

So, you've done something that improved it.

**Comment by bandit-ibayashi** [comment]
@joarwilk Thank you for sharing the latest updates. I'm glad to hear that this issue has been resolved in the latest version for iOS. It’s possible that the improvements made in previous updates have contributed to this.

**Comment by bandit-ibayashi** [comment]
The report has been improved with the update, so we will close this issue.
If similar issues still occur, please create a new one.
